SB-1095 getting rid of the java.io.File from the web-core and other parts of the code

fuss86 opened the pull request.

sbespalov commented on strongbox-storage/strongbox-storage-core/src/main/java/org/carlspring/strongbox/io/ArtifactFile.java:
```diff
@@ -1,50 +0,0 @@
- * @author mtodorov
- * 
- */
-public class ArtifactFile
```
@fuss86 you are cool! I did not dare to do it in my time, but always wanted :)

carlspring replied:
Good job! :)

fuss86 replied:
:)

sbespalov commented on strongbox-web-core/src/main/java/org/carlspring/strongbox/controllers/nuget/NugetArtifactController.java:
```diff
@@ -681,26 +656,26 @@ private URI storePackage(String storageId,
                                  nupkgFile.getVersion());
 
-            nugetArtifactManagementService.validateAndStore(storageId, repositoryId, path, new FileInputStream(hashFile));
+            nugetArtifactManagementService.validateAndStore(storageId, repositoryId, path, Files.newInputStream(hashFile));
```
@fuss86 many thanks for this too :smiley:

carlspring replied:
+1

fuss86 replied:
:)

carlspring commented on strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout-provider/src/test/java/org/carlspring/strongbox/cron/jobs/RegenerateMavenChecksumCronJobTestIT.java:
```diff
@@ -145,22 +145,22 @@ public void testRegenerateArtifactChecksum()
-        FileUtils.deleteIfExists(new File(snapshotArtifact_1.getFile()
-                                                            .getAbsolutePath()
-                                                            .replaceAll("jar", "pom") + ".sha1"));
+        TestFileUtils.deleteIfExists(new File(snapshotArtifact_1.getFile()
```
This looks good, but I think it would be even better to just do a `static import` for the necessary method. This way there'll be less unnecessary code to go through.

carlspring commented on strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-npm-layout-provider/src/main/java/org/carlspring/strongbox/repository/NpmRepositoryManagementStrategy.java:
```diff
@@ -1,43 +1,10 @@
+        extends AbstractRepositoryManagementStrategy
 {
 
-    private static final Logger logger = LoggerFactory.getLogger(NpmRepositoryManagementStrategy.class);
```
Nice cleanup! :)

carlspring commented on strongbox-web-core/src/main/java/org/carlspring/strongbox/controllers/nuget/NugetArtifactController.java:
```diff
@@ -459,7 +434,7 @@ public ResponseEntity putPackage(@RequestHeader(name = "X-NuGet-ApiKey", require
         {
             ServletInputStream is = request.getInputStream();
-            FileInputStream packagePartInputStream = extractPackageMultipartStream(extractBoundary(contentType), is);
+            InputStream packagePartInputStream = extractPackageMultipartStream(extractBoundary(contentType), is);
```
+1: Yeah, we should use interfaces, instead of implementations (wherever possible)! :)

carlspring commented:
This pull looks great! I have some minor remarks and would be happy to merge it afterwards! :)

Good job! ;)

fuss86 commented:
Please remember that it's a first part of the cleanup.

carlspring commented:
@fuss86,

I see you have removed the `ArtifactClient` and I have some comments and remarks:
* This is actually used. Initially, the idea was to have this used for test cases, but then I thought we should also have this sort of client available to anyone wanting to use the REST API, so that they shouldn't have to code things themselves, as we already have this.
* The `ArtifactClient` was mainly focused on the Maven side of things and was useful back in the days when we weren't using rest-assured for our testing.
* The `ArtifactClient` is currently used in the [`strongbox-web-integration-tests`](https://github.com/strongbox/strongbox-web-integration-tests/), (for example, [here](https://github.com/strongbox/strongbox-web-integration-tests/blob/master/maven/src/it/deploy-and-delete-artifact-directory-to-trash/clean.groovy#L8)). If I were to merge your pull as it is at the moment, it would break these tests. They're only a handful, but they will all break, so we would have to fix things there as well.
* The `ArtifactClient` will have to be renamed to `MavenArtifactClient` and we should have versions of this class per layout provider (at some point), due to the different needs that will arise for those specific implementations.
* If the `ArtifactClient` class is dropped, then the `IArtifactClient` interface should be renamed to `ArtifactClient` (I never liked the name of this interface).

carlspring commented on strongbox-storage/strongbox-storage-api/src/test/java/org/carlspring/strongbox/artifact/coordinates/MockedMavenArtifactCoordinates.java:
```diff
@@ -49,45 +49,6 @@ public MockedMavenArtifactCoordinates()
-        this(ArtifactUtils.convertPathToArtifact(path));
-    }
-
-    public MockedMavenArtifactCoordinates(String... coordinateValues)
```
I know this was ugly, but what was the to remove this constructor?

fuss86 replied:
Not used probably, will revert it back

carlspring commented on strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout-provider/src/main/java/org/carlspring/strongbox/artifact/MavenArtifact.java:
```diff
@@ -0,0 +1,32 @@
+    @Override
+    default File getFile()
+    {
+        throw new UnsupportedOperationException("Use getPath instead");
```
+1: This looks like a really good idea! :D

carlspring commented on strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout-provider/src/test/java/org/carlspring/strongbox/artifact/MockedMavenArtifactCoordinates.java:
```diff
@@ -0,0 +1,186 @@
+ */
+@XmlRootElement(name = "maven-artifact-coordinates")
+@XmlAccessorType(XmlAccessType.NONE)
+public class MockedMavenArtifactCoordinates
```
This "useless" class was wherever you moved it from, because it's sometimes required to test something with "dummy" coordinates in modules which are generic and have no actual implementation of `ArtifactCoordinates` exposed to them via a dependency, as it would cause a circular dependency.

It's not great, but it does the job for testing purposes. I would suggest restoring it to it's original location, as the need for something like this occurs every now and then.

fuss86 replied:
will revert it back to the previous state :)

carlspring replied:
Thanks! :)

carlspring commented on strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout-provider/src/test/java/org/carlspring/strongbox/cron/jobs/RegenerateMavenChecksumCronJobTestIT.java:
```diff
@@ -225,22 +223,22 @@ public void testRegenerateChecksumInRepository()
-                                                            .getAbsolutePath()
-                                                            .replaceAll("jar", "pom") + ".sha1"));
+        deleteIfExists(new File(snapshotArtifact_2.getPath()
+                                                                .toString() + ".md5"));
```
This could be a one-liner (the same is true in the cases below).

carlspring commented on strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout-provider/src/test/java/org/carlspring/strongbox/cron/jobs/RegenerateMavenChecksumCronJobTestIT.java:
```diff
@@ -249,26 +247,26 @@ public void testRegenerateChecksumInRepository()
                     assertTrue("The checksum file for artifact doesn't exist!",
-                               new File(snapshotArtifact_2.getFile()
-                                                          .getAbsolutePath() + ".sha1").exists());
+                               new File(snapshotArtifact_2.getPath()
```
Could be one-liners (here and below).

carlspring commented on strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout-provider/src/test/java/org/carlspring/strongbox/testing/TestCaseWithMavenArtifactGeneration.java:
```diff
@@ -236,7 +237,7 @@ public Artifact createTimestampedSnapshotArtifact(String repositoryBasedir,
 
-    public Artifact createTimestampedSnapshot(String repositoryBasedir,
+    public MavenArtifact createTimestampedSnapshot(String repositoryBasedir,
                                               String groupId,
```
Cosmetic: The parameters are no longer aligned.

carlspring commented on strongbox-web-core/src/test/java/org/carlspring/strongbox/controllers/maven/MavenArtifactControllerTest.java:
```diff
@@ -860,15 +873,15 @@ public void testMetadataAtGroupAndArtifactIdLevel()
-        Metadata groupLevelMetadata = client.retrieveMetadata("storages/" + STORAGE0 + "/" +
+        Metadata groupLevelMetadata = defaultMavenArtifactDeployer.retrieveMetadata("storages/" + STORAGE0 + "/" +
                                                               REPOSITORY_RELEASES2 + "/" +
                                                               ArtifactUtils.getGroupLevelMetadataPath(artifact1));
```
Cosmetic: The method parameters are no longer aligned (here and below).

carlspring commented on strongbox-web-core/src/test/java/org/carlspring/strongbox/rest/common/MavenRestAssuredBaseTest.java:
```diff
@@ -152,13 +137,13 @@ public String createSnapshotVersion(String baseSnapshotVersion,
+                                                                    String baseSnapshotVersion,
+                                                                    String packaging,
+                                                                    String[] classifiers,
+                                                                    int numberOfBuilds)
```
Cosmetic: The method parameters are no longer aligned.

carlspring commented:
@fuss86,

I've reviewed this version of the PR (the first part of our larger refactoring to kill the use of `File`-s). It looks great!

I have a few remarks, about some side-effects, as we discussed and I won't be merging this until you have restored the artifact client classes (in whatever form you find fit), as it will break the `strongbox-web-integration-tests` that we've been working so hard to stabilize). But... other than that, this pull will bring us one step closer to our goal!

Thanks for hard work! :)

fuss86 commented:
@carlspring , referring to `ArtifactClient` - it has been reverted back without name change. However I moved this class to the `strongbox-testing-storage` module (as well as `BaseArtifactClient` and `IArtifactClient`) - these are all classes used for test purposes only. The package name didn't change. 

I've also found two classes used only for test purposes in the `strongbox-rest-client` module. I moved them to the  `strongbox-testing-storage` too, which made `strongbox-rest-client` module unused and unnecessary, so I got rid of it.

Summarizing:

From  `strongbox-client` to  `strongbox-testing-storage` 
* `ArtifactClient` 
* `BaseArtifactClient`
* `IArtifactClient`


From  `strongbox-rest-client` to  `strongbox-testing-storage` 
* `RestClient` 
* `RestAssuredArtifactClient` 

Now we will know that they are usable only for tests.
